common/frame.go

```diff
@@ -518,12 +518,14 @@ func (f *Frame) setContext(world executionWorld, execCtx frameExecutionContext)
 	if f.executionContexts[world] != nil {
 		f.log.Debugf("Frame:setContext", "fid:%s furl:%q ectxid:%d world:%s, world exists",
 			f.ID(), f.URL(), execCtx.ID(), world)
-		return
+		return errors.New("context already set")
```

Comment:
Have you considered the suggestion in https://github.com/grafana/xk6-browser/issues/482#issuecomment-1219385591 to remove this check and always overwrite the context with the newly provided one?

Or is this guarding the case when we're processing a second `setContext()` call without having processed the first `onExecutionContextsCleared()`? Wouldn't we "lose" this context in that case? :thinking:

Reply:
Since the issue is due to out of order events, it didn't feel correct to then change the behaviour in `setContext`, and i also don't know if it's possible to have a `setContext` without first having a `clearContext` after the initial navigation to `about:blank`.

As you've pointed out, this is currently helping to identify when the context has been set, but needs to be cleared (assumed that this must happen first). We shouldn't lose the context as long as the clear does take place before setting it again.

We need a sensible timeout while waiting for the (out of order) clear context event, but there's a small chance that the clear context event is delayed more than 1-5 secs or is missing. In that case:

- should we overwrite the context still, but a late clear context event will clear the context resulting in the same deadlock?
- Or should we leave the behaviour as is, but a late clear context event will clear the context resulting in the same deadlock?
- Or should we panic and exit since we haven't seen the expected clear context event and avoid the deadlock?

Reply:
Hhmm right then it would make sense to return an error here, since the situation of overwriting contexts is a bug. I think panicking would be an extreme approach to take, and wouldn't solve the issue, but returning an error is fine in this case. Thanks for explaining.